### PR TITLE
[DeadCode] Skip Remove @return/@param null functionality in case of sub type

### PR DESCRIPTION
**Pull request by samsonasik**
@TomasVotruba PR:

- https://github.com/rectorphp/rector-src/pull/5348

cause a lot of invalid usage, as they checked too early, even when description exists, so it make remove `@return/@param` null only when typed already.

### CODE REVIEW

**Comment by samsonasik** [comment]
@TomasVotruba this only remove `@param null/@return null` only when:

- They are typed native already
- They don't have description

**Comment by TomasVotruba** [comment]
> They are typed native already

Native type is not relevant.


> They don't have description

This is indeed valid. Could you add this one?

**Comment by TomasVotruba** [comment]
What could be done to make fixing errors easy is to have a separated RemoveNullTagValueNodeRector rule, which would handle exactly these cases. 
What do you think?

**Comment by samsonasik** [comment]
I added param null with description fixture https://github.com/rectorphp/rector-src/pull/5350/commits/349cd91e18a1e4eb509f21a03a111c220413fe57, it previously removed on your PR  because of too early remove.

File: rules-tests/DeadCode/Rector/ClassMethod/RemoveUselessReturnTagRector/Fixture/remove_null.php.inc [hunk 1]
```diff
@@ -7,7 +7,7 @@ class RemoveNull
     /**
      * @return null
      */
-    function foo()
+    function foo(): ?int
     {
         return null;
     }
```
Comment (TomasVotruba) [inline rules-tests/DeadCode/Rector/ClassMethod/RemoveUselessReturnTagRector/Fixture/remove_null.php.inc]: These should be kept

Comment (samsonasik) [reply rules-tests/DeadCode/Rector/ClassMethod/RemoveUselessReturnTagRector/Fixture/remove_null.php.inc]: sure https://github.com/rectorphp/rector-src/pull/5350/commits/ede7425008c805852fbd532a12392887df662f8e,

Comment (TomasVotruba) [reply rules-tests/DeadCode/Rector/ClassMethod/RemoveUselessReturnTagRector/Fixture/remove_null.php.inc]: I mean this fixture file should be untouched.

**Comment by samsonasik** [comment]
@TomasVotruba I think we need to revert your PR first https://github.com/rectorphp/rector-src/pull/5348 , then we can discuss how it should be applied.

The rule description is: "Remove @return docblock with same type as defined in PHP"

File: rules-tests/DeadCode/Rector/ClassMethod/RemoveUselessReturnTagRector/Fixture/remove_null.php.inc (hunk 1 above)
Comment (samsonasik) [reply rules-tests/DeadCode/Rector/ClassMethod/RemoveUselessReturnTagRector/Fixture/remove_null.php.inc]: The reason why `@return null` actually valid here, is that it sub type of defined type, which make more correct result of what user want to.

The typed can be enforced by interface, so can't change, but the only way to enforce it is via `@return null`.

Comment (samsonasik) [reply rules-tests/DeadCode/Rector/ClassMethod/RemoveUselessReturnTagRector/Fixture/remove_null.php.inc]: Remove `null` only make sense when it exactly different type ( which already handled ), eg:

```php
/** @return null */
function foo(): int
```

**Comment by samsonasik** [comment]
I propose to revert your PR first:

- https://github.com/rectorphp/rector-src/pull/5351

then we can re-arrange the way `null` docblock should be applied, eg as you suggest, via separate rule: `RemoveNullTagValueNodeRector`

**Comment by TomasVotruba** [comment]
I prefer to have moving functionality to the new rule instead of removing then adding.
Could you create the separate rule in the one PR?

**Comment by samsonasik** [comment]
That mean the remove `@param/@return null` you did should be reverted first as works on sub type. https://github.com/rectorphp/rector-src/pull/5350/commits/ede7425008c805852fbd532a12392887df662f8e and https://github.com/rectorphp/rector-src/pull/5350/commits/d685745a634c135c7d53cad0e7547e966cd43760

Do you mean this PR can be merged fist ?

**Comment by TomasVotruba** [comment]
Not reverting, but moving the logic to new `RemoveNullTagValueNodeRector` rule. Both in this PR.
Part of dead code set.

**Comment by samsonasik** [comment]
What I mean is: current functionality became incorrect because too early remove null, as it also remove `@return null description`, so should be reverted.

The remve null functionality can be moved with better logic, but this need to be reverted first, then we can focus on new rule.

**Comment by samsonasik** [comment]
Remove null same type already covered, see fixture https://github.com/rectorphp/rector-src/pull/5350/commits/b5c8dc7baf91605bc70807cb32f6bc8d80f837fa

**Comment by TomasVotruba** [comment]
Could you open the PR for alternative rule first?

**Comment by samsonasik** [comment]
in this PR as well ?

**Comment by TomasVotruba** [comment]
Up to you :+1:

**Comment by samsonasik** [comment]
I think new rule should not be part of `config/dead-code.php` as directly, but can be manually added as "feature" to avoid on purspose have null tag.

I will create separate PR, just don't create a tag first on current dev-main, as it will cause catastrophic in our project.

**Comment by samsonasik** [comment]
I created new PR for remove null tag:

- https://github.com/rectorphp/rector-src/pull/5352

**Comment by TomasVotruba** [comment]
Sure 👍
No tagging planned, nearest in about a week or two.

**Comment by TomasVotruba** [comment]
Thanks 🙏

**Pull request merged**